### ISSUE TITLE
Adaptive Sharpness Aware Minimization (ASAM) Implementation Help

### ISSUE THREAD

**Issue by evanatyourservice** [davda54/sam#22]
Hello! I've been getting very good results using SAM. I stumbled on this paper, [ASAM: Adaptive Sharpness-Aware Minimization for Scale-Invariant Learning of Deep Neural Networks](https://arxiv.org/abs/2102.11600), from Samsung Research improving upon SAM. ASAM is SAM but scale invariant, so it minimizes a relative neighborhood of weights vs. a perfect circle of weights and further improves generalization. Would you like to help me implement it?

#### This is what I have so far but I am not sure it is correct

They add an operator, T_w, described on page 8 in their paper, into the epsilon equation in a couple places. I am using their element-wise T_w. They show their epsilon equation on page 11 under where it says "for t = 0, 1, 2, · · · . Especially, if p = 2,", multiplying the grads by T_w before the norm calc, and multiplying the epsilon numerator by T_w squared.

This is my thought for the norm calc. I am not sure if I have it right, if I need to do something with torch.diag or not.

```
        norm = torch.norm(
                    torch.stack([
                        (torch.abs(p) * p.grad).norm(p=2).to(shared_device)
                        for group in self.param_groups for p in group["params"]
                        if p.grad is not None
                    ]),
                    p=2
               )
```

And this is my thought for epsilon, but I again am not sure if I need to do something with torch.diag, or need to flatten and concat things in a certain way or not.

```
e_w = torch.pow(p, 2) * p.grad * scale.to(p)
```

Any thoughts? I'm sorry if you are busy, but figured I'd ask! I've only tested on mnist so far, and did get a slightly improved result vs. SAM, so I think this is at least on the right track if not correct.

**Comment by guozhiyao** [comment]
Hi, is the hyperparameter `rho` used in your SAM set to 0.05?

**Comment by evanatyourservice** [comment]
The default for SAM from the authors is 0.05.

**Comment by guozhiyao** [comment]
In my face recognition model, the default `rho` will damage the performance

**Comment by davda54** [comment]
Hi, thank you very much for showing us this awesome paper! It seems that the results are strictly better than SAM's, so it would be a nice addition to this repository. Have you tried your implementation on other datasets than MNIST?

To be honest, I'm a little confused with the definition of `T_w` in equation (4), it seems to be unnecessarily convoluted. That being said, I believe your implementation is correct -- at least it makes sense to me :)

I will try to test ASAM on Cifar10 and then add it to this repository.

**Comment by evanatyourservice** [comment]
Awesome sounds good. No I haven't tried any other tests yet, I'll try the ones from their paper too to see if I get the same results as their paper describes. I reached out to the authors with a couple questions but they haven't gotten back to me.

**Comment by davda54** [comment]
I tried ASAM with `rho=0.5` on a simple WRN and it makes the performance only marginally better. Of course, that may be caused by the smallish model I used.

Anyway, good news is that the performance is not worse, so your implementation seems to work :) I've added the element-wise ASAM to this repository, thank you so much for your contribution!

**Comment by evanatyourservice** [comment]
Excellent. Your results look similar to the improvements shown in the ASAM paper -- they're improvement on SAM is marginal vs. SAM's improvement on the regular results, like 1% better or so. I guess that's a good sign, then, maybe this implementation is correct. Good addition because I suppose it further prevents overfitting when using more epochs.

Thanks for the shoutout in the readme :)

**Comment by bonlime** [comment]
Hi,
I've encountered SAM paper today and after that found this implementation. After carefully reading the paper there is one question which still confuses me. The equation (2) for computing epsilon_hat looks like this and uses norm over all gradients in the model:
![image](https://user-images.githubusercontent.com/14337581/128206774-9074deaa-b2ca-443d-bc19-28ae642b1519.png)

In your code this is implemented in https://github.com/davda54/sam/blob/d212252f0ecc3b2e9fa4198e0606d7fb06907b2c/sam.py#L48

But from Figure 2 it looks like that each weight should use norm of its' own gradient only. This also makes more sense to me, because if we calculate norm of all gradients, large gradients from few parameters would reduce epsilon for all parameters which is undesired
![image](https://user-images.githubusercontent.com/14337581/128207707-fd71deb2-f178-42f6-aca6-1f69133918b7.png)
 
What do you think about it?

**Comment by davda54** [comment]
Hi, good idea! Using a separate norm for each layer seems reasonable, especially for bigger models. You should definitely try it! :)
